# **Loan Performance Analysis**

1. [Introduction](#Introduction)
2. [Data Preparation (SQL)](#Data-Preparation-(SQL))
3. [Analyzing Financial Loan Data in SQL](#Analyzing-Financial-Loan-Data-in-SQL)
4. [Creating SQL Tables for Dashboard Overview](#Creating-SQL-Tables-for-Dashboard-Overview)
5. [Tableau Dashboard](#Tableau-Dashboard)
6. [Detailed Analysis](#Detailed-Analysis)
7. [Conclusion](#Conclusion)

## **Introduction**

### **Background**

In 2021, as society began to emerge from the grips of the COVID-19 pandemic, the US economy faced significant challenges. Business activity had slowed, isolation and restrictions had disrupted normal life, and the financial stability of many individuals and businesses was precarious. To combat these issues and stimulate recovery, the US government implemented measures such as lowering interest rates to make loans more affordable and accessible.

The goal of these lower interest rates was to inject much-needed liquidity into the economy, encouraging both individuals and businesses to borrow and spend. This increased spending and investment was intended to help businesses recover and potentially boost employment, which had been negatively impacted by the pandemic's economic slowdown.

### **Objective**

The primary objective of this project is to conduct a comprehensive analysis of loan records issued during the critical year of 2021. By delving into this dataset, we aim to gain a deep understanding of how loans performed during this period, characterized by the gradual recovery of the economy from the COVID-19 pandemic and the government's implementation of policies designed to stimulate borrowing and spending.

Our investigation will specifically focus on evaluating the impact of the government's initiative to encourage lower interest rates. We will assess how these reduced rates influenced both the behaviour of borrowers and the ultimate outcomes of the loans themselves. This analysis will provide valuable insights into the effectiveness of this policy in fostering economic recovery, as well as its implications for both lenders and borrowers within the unique financial landscape of the time.

### **Data Source**

The dataset used for this project offers a comprehensive overview of lending activities across the United States during the pivotal year of 2021. Comprising loan records from various banks nationwide, this dataset provides a unique snapshot of the financial landscape during a period marked by economic recovery and government initiatives to stimulate borrowing.

Each loan record within the dataset contains a wealth of information, encompassing both borrower-specific details and loan-specific attributes. This includes crucial borrower information such as employment status, income verification, and other relevant financial indicators that banks typically consider when assessing loan applications. Additionally, the dataset details the specifics of each loan, including the loan amount, interest rate, current payment status (e.g., current, paid off), and any adverse outcomes such as default or charge-off.

To provide a clearer understanding of the dataset's structure and content, we present a breakdown of the key columns and their respective meanings:


| Column                 | Description                                                                                                                          | Type    | Classification |
|------------------------|--------------------------------------------------------------------------------------------------------------------------------------|---------|----------------|
| `id`                   | **Loan ID:** Unique identifier for each loan, serving as a primary key to track and manage loans throughout their lifecycle.         | Integer | Numerical      |
| `address_state`        | **Address State:** Indicates the borrower's location to assess regional risks and compliance with state regulations.                 | String  | Categorical    |
| `application_type`     | **Application Type:** Type of application (Individual or Joint), differentiating the processing and requirements.                    | String  | Categorical    |
| `emp_length`           | **Employee Length:** Provides insights into employment stability, indicating job security and repayment ability.                    | String  | Categorical    |
| `emp_title`            | **Employee Title:** Specifies the borrower's job title, aiding in income source verification and financial capacity assessment.    | String  | Categorical    |
| `grade`                | **Grade:** Risk classification assigned based on creditworthiness, influencing loan pricing and attractiveness to investors.         | String  | Categorical    |
| `home_ownership`       | **Home Ownership:** Indicates housing status, providing insights into financial stability and collateral availability.               | String  | Categorical    |
| `issue_date`           | **Issue Date:** Marks the origination date of the loan, crucial for tracking, interest calculations, and maturity assessments.       | String  | Categorical    |
| `last_credit_pull_date`| **Last Credit Pull Date:** Records when the borrower’s credit report was last accessed, helping monitor creditworthiness updates.   | String  | Categorical    |
| `last_payment_date`    | **Last Payment Date:** Indicates the most recent payment received, tracking payment history and assessing delinquency.              | String  | Categorical    |
| `loan_status`          | **Loan Status:** Shows the current state of the loan (e.g., fully paid, current, default), used for risk analysis and provisioning. | String  | Categorical    |
| `next_payment_date`    | **Next Payment Date:** Estimates the date of the next loan payment, aiding in cash flow forecasting and revenue projection.         | String  | Categorical    |
| `member_id`            | **Member ID:** Unique identifier for the member, used to track individual borrowers across multiple loans.                           | Integer | Numerical      |
| `purpose`              | **Purpose:** Specifies the reason for the loan (e.g., debt consolidation, education), aligning loan terms with borrower needs.       | String  | Categorical    |
| `sub_grade`            | **Sub Grade:** Provides a finer level of risk differentiation within grades, tailoring terms and interest rates to risk profiles.    | String  | Categorical    |
| `term`                 | **Term:** Defines the loan's duration in months, setting the repayment period and structuring loan agreements.                       | String  | Categorical    |
| `verification_status`  | **Verification Status:** Indicates the verification of borrower's financial information, assessing data accuracy and credibility.    | String  | Categorical    |
| `annual_income`        | **Annual Income:** Total yearly earnings of the borrower, used to determine loan eligibility and evaluate creditworthiness.          | Float   | Numerical      |
| `dti`                  | **DTI (Debt-to-Income Ratio):** Measures debt burden relative to income, assessing capacity for additional debt.                     | Float   | Numerical      |
| `installment`          | **Instalment:** Fixed monthly payment amount for the loan, used to structure terms and assess payment affordability.                 | Float   | Numerical      |
| `int_rate`             | **Interest Rate:** Annual cost of borrowing, key in loan pricing, profit management, and investor attraction.                        | Float   | Numerical      |
| `loan_amount`          | **Loan Amount:** Total sum borrowed, defining the principal amount, used to determine the size and scope of the loan.                | Integer | Numerical      |
| `total_acc`            | **Total Accounts:** Total number of credit lines in the borrower's credit file, used for comprehensive credit assessment.            | Integer | Numerical      |
| `total_payment`        | **Total Payment:** Payments received to date for total amount funded, important for financial tracking and analysis.                 | Integer | Numerical      |

### **Analysis Goals**

To achieve our objectives, our analysis will focus on the following goals:

* **Identify key factors influencing loan outcomes:** Explore the relationship between loan performance and factors such as loan purpose, verification status, and geographic location.
* **Quantify the financial impact:** Calculate the financial gains and losses experienced by lenders due to successful loan repayments and defaults.
* **Uncover patterns and trends:** Identify any recurring patterns or trends in loan behaviour, particularly those related to borrower characteristics, loan types, or regional differences.
* **Inform potential policy adjustments:** Assess whether further modifications to interest rates or lending practices could encourage greater borrowing and stimulate economic growth.

By addressing these analysis goals, we will gain a comprehensive understanding of the lending landscape in 2021 and its implications for both borrowers and lenders in the unique economic context of the post-pandemic recovery period.

## **Data Preparation in SQL**

### **Initializing the Database Environment**

**In order to prepare the environment for data analysis, I first established a database within the SQL Server environment. The following steps outline the process I took to create the database for this project:**

**1. Connecting to SQL Server Management Studio (SSMS)**

* I started by opening Microsoft SQL Server Management Studio (SSMS).
* Then, I entered the appropriate server name (or IP address) and my login credentials to connect to the SQL Server instance.

**2. Creating a New Database**

* In the Object Explorer, I right-clicked on the "Databases" folder and chose "New Database".
* I named the new database `"Bank Loan DB"` to keep it easily identifiable.
* For this initial setup, I left the remaining options as their defaults and clicked "OK" to finish creating the database.

![1. new db.png](../src/img/SQL/Data_Preparation/1_new_db.png)

![2. new db.png](../src/img/SQL/Data_Preparation/2_new_db.png)

**3. Verifying Database Creation**

* To make sure everything worked, I refreshed the "Databases" folder in Object Explorer. My newly created `"Bank Loan DB"` database was there as expected.

### **Importing the Financial Loan Data**

**1. Importing the CSV Data**

After successfully creating the database, I proceeded to import the "financial_loan.csv" file. Within SQL Server Management Studio (SSMS), I navigated to the newly created database and right-clicked on the "Tables" folder. From the context menu, I selected "Import Flat File" to initiate the import process. 

![4. new db.png](../src/img/SQL/Data_Preparation/3_new_db.png)

**2. Specifying File Location and Table Name**

* In the "Import Flat File" wizard, I located the "financial_loan.csv" file on my computer and provided its path.
* To store the imported data within the database, I assigned the name `"bank_loan_data"` to the new table that would be created. 

**3. Refining the Table Schema for Data Accuracy and Efficiency**

After importing the data, I carefully reviewed the table schema generated by SSMS. To optimize the table's structure and ensure data accuracy, I made several adjustments:

**3.1. Primary Keys**

* **What is a Primary Key:** A primary key is a column (or a combination of columns) that serves as a unique identifier for every row within a table. Primary keys are essential for maintaining data integrity, ensuring that each record can be reliably located and referenced.

* **Our Choice:** In this case, I designated the "id" column as the primary key, recognizing that it holds unique values for each loan record.

**3.2. Choosing between VARCHAR and NVARCHAR**

* **Rationale:** I noticed that many string columns were initially assigned the NVARCHAR(50) datatype. Knowing that our dataset exclusively uses English characters, I converted these to VARCHAR(50) to improve storage efficiency. NVARCHAR stores data using Unicode, which supports a wider range of characters, while VARCHAR uses a more space-efficient encoding for standard English text.

**3.3. Adjusting Character Limits:**

* **Example:** For the "emp_title" column, I increased the character limit from 50 to 100 to accommodate some longer job titles present in the data.

**3.4. Changing INT vs. SMALLINT**

* **Reasoning:** I observed that the "loan_amnt" and "total_pymnt" columns were initially set as SMALLINT. Since our dataset contains loan values that exceed the maximum range supported by SMALLINT, I changed the datatype to INT to prevent potential errors. INT allows for a wider range of integer values.

By refining the table schema, I was able to successfully import the CSV file into the database with the appropriate data types and constraints in place.

**BEFORE:**

![6. new db before.png](../src/img/SQL/Data_Preparation/5_new_db_before.png)

**AFTER:**

![6. new db afterpng.png](../src/img/SQL/Data_Preparation/5_new_db_after.png)

**4. Verifying Successful Import**

To ensure the CSV data was imported correctly, I opened a new query window within SSMS. Using the following query, I was able to view the entire contents of the newly created table, confirming successful data import:

![7. new db 1.png](../src/img/SQL/Data_Preparation/6_new_db_1.png)

```sql
SELECT 
    * 
FROM 
    bank_loan_data 
```

I carefully checked that the query results matched the original CSV data in terms of column names, number of rows, and overall structure. This verification confirmed that the import process was successful and I was ready to move on to the next phase of my project.

![7. new db 2.png](../src/img/SQL/Data_Preparation/6_new_db_2.png)

## **Analyzing Financial Loan Data in SQL**

**Loan Performance Report and Key Performance Indicators (KPIs)**

This section focuses on leveraging the imported loan data to create a comprehensive loan performance report. The report will provide valuable insights into key metrics associated with the bank's lending activities. By analyzing these metrics, I can gain a deeper understanding of the bank's portfolio health, identify potential trends, and ultimately support data-driven decision-making.

**Utilizing KPIs for Informed Decisions**

To achieve these goals, I will utilize a set of Key Performance Indicators (KPIs) specifically tailored to the loan dataset. These KPIs act as quantifiable measures that provide a clear picture of the bank's lending performance. By calculating and analyzing these KPIs, I can gain valuable insights into areas such as loan profitability, risk management, and overall portfolio health.

**Understanding the Data Through KPIs**

By systematically calculating these KPIs and interpreting the results, I can transform the raw data into actionable insights. This information will empower stakeholders to make informed decisions regarding lending strategies, risk mitigation tactics, and overall portfolio optimization. 

This approach allows us to move beyond simply "looking at the data" and instead leverage its power to gain a deeper understanding of the bank's lending activities and their impact on financial performance.

### **Summary Key Performance Indicators (KPIs) Analysis**

Here's the outline of the KPIs I'll explore within this section:

#### **1. Total Loan Applications:**

**Explanation:** This KPI provides insights into the overall volume of loan applications received by the bank. Analyzing changes in this KPI over time can reveal trends in borrowing demand and help assess the efficacy of the bank's marketing and outreach efforts.

**Components:**

1.  **Total Number of Loans:**
    *   **Explanation:** This metric simply represents the total number of loan applications present within the dataset, reflecting the overall volume of lending activity for the bank.
    *   **Query:**

        ```sql
        SELECT
            COUNT(*) AS Total_Loan_Applications
        FROM
            bank_loan_data;
        ```
    *   **Results:**
    
        ![Total Loan Applications.png](../src/img/SQL/KPI/1.%20Total%20Loan%20Applications/Total_Loan_Applications.png)

2.  **Number of Loans by Month:**
    *   **Explanation:** Since the data is from 2021, it makes more sense to analyze loan volume on a monthly basis instead of the typical "month to date" approach. Tracking loan application volume by month helps identify seasonal trends or shifts in lending activity throughout the year.
    *   **Query:**

        ```sql
        SELECT
	        DATENAME(month, issue_date) AS Month,
	        COUNT(*) AS Loan_Application_Count
        FROM
	        bank_loan_data
        GROUP BY
	        DATENAME(month, issue_date),
	        MONTH(issue_date)
        ORDER BY
            MONTH(issue_date);
        ```
    *   **Results:**

        ![mtd loan count.png](../src/img/SQL/KPI/1.%20Total%20Loan%20Applications/mtd_loan_count.png)

    *   **Note:** To simplify the next calculation, I've stored the revised query results in a new table called "monthly_loan_counts"
    *   **Query:**

        ```sql
        SELECT
            DATENAME(month, issue_date) AS Month,
            MONTH(issue_date) AS month_num,
            COUNT(*) AS Loan_Application_Count
        INTO
            monthly_loan_counts
        FROM
            bank_loan_data
        GROUP BY
            DATENAME(month, issue_date),
            MONTH(issue_date)
        ORDER BY
            MONTH(issue_date);
        ```

3.  **Month-Over-Month (MoM) Percentage Change in Number of Loans**
    *   **Explanation:** This metric calculates the percentage change in loan applications from one month to the next. Monitoring this change can highlight periods of accelerated or decreased lending activity and indicate potential shifts in market conditions or internal lending policies.
    *   **Query:**

        ```sql
        SELECT 
	        m1.Month,
	        CAST(((m1.Loan_Application_Count - m2.Loan_Application_Count) * 100.0 / m2.Loan_Application_Count) AS DECIMAL(10,3)) AS Loan_MoM_Change_Percent
        FROM 
	        monthly_loan_counts AS m1
        LEFT JOIN 
            monthly_loan_counts AS m2
        ON 
	        m1.month_num = m2.month_num + 1
        ORDER BY
	    m1.month_num;
        ```
    *   **Results:**
  
        ![mom num loan pct.png](../src/img/SQL/KPI/1.%20Total%20Loan%20Applications/mom_num_loan_pct.png)

#### **2. Total Funded Amount:**

**Explanation:** This KPI reveals the total dollar amount distributed by the bank through loans within a given period. Analyzing this metric helps understand the bank's overall lending volume and can shed light on changes in lending activity and risk tolerance over time.

**Components:**

1. **Total Funded Amount:**
   * **Explanation:** This metric calculates the overall sum of loans funded by the bank within the dataset.
   * **Query:** 
      ```sql
      SELECT
        SUM(loan_amount) AS total_funded_amount
      FROM
        bank_loan_data;
      ```
    * **Results:**

      ![image.png](../src/img/SQL/KPI/2.%20Total%20Funded%20Amount/total.png)

2. **Total Funded Amount by Month:**
   * **Explanation:** Since the data is from 2021, it makes sense to track the total funded amount on a monthly basis instead of a traditional "month to date" approach. This allows to identify potential seasonal fluctuations or other trends in the bank's lending volume throughout the year.
   * **Query:** 
      ```sql
      SELECT
         DATENAME(month, issue_date) AS month,
         SUM(loan_amount) AS total_funded_amount
      FROM
         bank_loan_data
      GROUP BY
         DATENAME(month, issue_date),
         MONTH(issue_date)
      ORDER BY
         MONTH(issue_date);
      ```
    * **Results:**

      ![image.png](../src/img/SQL/KPI/2.%20Total%20Funded%20Amount/mtd.png)

    *   **Note:** To simplify the next calculation, I've stored the revised query results in a new table called "monthly_total_fund".
    *   **Query:**

        ```sql
        SELECT
            DATENAME(month, issue_date) AS month,
            MONTH(issue_date) AS month_num,
            SUM(loan_amount) AS total_funded_amount
        INTO
            monthly_total_fund
        FROM
            bank_loan_data
        GROUP BY
            DATENAME(month, issue_date),
            MONTH(issue_date)
        ORDER BY
            MONTH(issue_date);
        ```

3. **Month-Over-Month (MoM) Percentage Change in Total Funded Amount:**
   * **Explanation:** This metric helps track the percentage change in the total dollar amount funded as loans from one month to the next. Analyzing this percentage change can reveal significant shifts in lending volume or changes in the target size of loans being approved.
   * **Query:** 
      ```sql
      SELECT 
         m1.Month,
         CAST(((m1.total_funded_amount - m2.total_funded_amount) * 100.0 / m2.total_funded_amount) AS DECIMAL(10,3)) AS total_funded_MoM_change_pct
      FROM 
         monthly_total_fund AS m1
      LEFT JOIN 
         monthly_total_fund AS m2
      ON 
         m1.month_num = m2.month_num + 1
      ORDER BY
         m1.month_num;
      ```
    * **Results:**

      ![image.png](../src/img/SQL/KPI/2.%20Total%20Funded%20Amount/mom.png)

#### **3. Total Amount Received**

**Explanation:** This KPI tracks the total sum of repayments received by the bank from borrowers. Analyzing this metric is crucial for assessing the bank's cash flow health and its ability to collect on outstanding loans. It can also indicate trends in borrower repayment behavior and potential delinquency rates.

**Components:**

1. **Total Amount Received:**
   * **Explanation:** This metric calculates the total sum of all loan repayments received by the bank within the dataset.
   * **Query:** 
      ```sql
      SELECT 
         SUM(total_payment) AS total_amount_received 
      FROM 
         bank_loan_data;
      ``` 
   * **Result:**
  
      ![image.png](../src/img/SQL/KPI/3.%20Total%20Amount%20Received/total_amount_recivied.png)

2. **Total Amount Received by Month:**
   * **Explanation:** Following the approach used in previous KPIs, instead of "month to date," I'll track the total amount received on a monthly basis. This allows to identify potential seasonal trends or patterns in borrower repayments throughout the year. 
   * **Query:** 
      ```sql
       SELECT
         DATENAME(month, issue_date) AS month,
         SUM(total_payment) AS total_funded_amount
      FROM
         bank_loan_data
      GROUP BY
         DATENAME(month, issue_date),
         MONTH(issue_date)
      ORDER BY
         MONTH(issue_date);
      ``` 
   * **Result:**

      ![image.png](../src/img/SQL/KPI/3.%20Total%20Amount%20Received/total_amount_recivied_mtd.png)

   * **Note:** To simplify the next calculation, I've stored the revised query results in a new table called "monthly_amount_received".
   * **Query:**
      ```sql
        SELECT
            DATENAME(month, issue_date) AS month,
            MONTH(issue_date) AS month_num,
            SUM(total_payment) AS total_amount_received
         INTO
            monthly_amount_received
         FROM
            bank_loan_data
         GROUP BY
            DATENAME(month, issue_date),
            MONTH(issue_date)
         ORDER BY
            MONTH(issue_date);
      ``` 

3. **Month-Over-Month (MoM) Percentage Change in Total Amount Received:**
   * **Explanation:** This metric helps track the percentage change in the total dollar amount received from loan repayments between one month and the next. Analyzing this percentage change can reveal trends in repayment patterns and potential areas for improvement in collection strategies.
   * **Query:** 
      ```sql
      SELECT
         m1.month AS Month,
         CAST(((m1.total_amount_received - m2.total_amount_received) * 100.0 / m2.total_amount_received) AS DECIMAL(10,3)) AS total_received_MoM_change_pct
      FROM
         monthly_amount_received AS m1
      LEFT JOIN
         monthly_amount_received AS m2
      ON
         m1.month_num = m2.month_num + 1
      ORDER BY
         m1.month_num;
      ``` 
   * **Result:**

      ![image.png](../src/img/SQL/KPI/3.%20Total%20Amount%20Received/total_amount_recivied_mom.png)

#### **4. Average Interest Rate**

**Explanation:** This KPI reveals the average interest rate charged by the bank across its loan portfolio. Tracking changes in this KPI over time can help assess shifts in the bank's pricing strategies, evolving market conditions, and potentially the risk profile of its lending activities.

**Components:**

1.  **Overall Average Interest Rate**
    *   **Explanation:** This metric calculates the average interest rate across all loans within the dataset.
    *   **Query:**
        ```sql
         SELECT
            ROUND(AVG(int_rate) * 100, 3) AS avg_interest_rate
        FROM
            bank_loan_data;
        ```
    *   **Result:**

        ![image.png](../src/img/SQL/KPI/4.%20Average%20Interest%20Rate/avg_intestest_rate_total.png)

2.  **Monthly Average Interest Rate**
    *   **Explanation:** This metric tracks the average interest rate on a month-by-month basis, revealing potential fluctuations in the bank's pricing strategies throughout the year.
    *   **Query:**
        ```sql
        SELECT
            DATENAME(month, issue_date) AS month,
            ROUND(AVG(int_rate) * 100, 3) AS avg_interest_rate
        FROM
            bank_loan_data
        GROUP BY
            DATENAME(month, issue_date),
            MONTH(issue_date)
        ORDER BY
            MONTH(issue_date);
        ```
    *   **Result:**

        ![image.png](../src/img/SQL/KPI/4.%20Average%20Interest%20Rate/avg_intestest_rate_mtd.png)
        
    * **Note:** To simplify the next calculation, I've stored the revised query results in a new table called "monthly_avg_interest".
    * **Query:**
        ```sql
        SELECT
            DATENAME(month, issue_date) AS month,
            MONTH(issue_date) AS month_num,
            ROUND(AVG(int_rate) * 100, 3) AS avg_interest_rate
        INTO
            monthly_avg_interest
        FROM
            bank_loan_data
        GROUP BY
            DATENAME(month, issue_date),
            MONTH(issue_date)
        ORDER BY
            MONTH(issue_date);
        ```

3.  **Month-Over-Month (MoM) Percentage Change in Average Interest Rate**
    *   **Explanation:** This metric calculates the percentage change in the average interest rate from one month to the next, highlighting shifts in the bank's interest rate levels over time.
    *   **Query:**
        ```sql
        SELECT
            m1.month AS Month,
            CAST(((m1.avg_interest_rate - m2.avg_interest_rate) * 100.0 / m2.avg_interest_rate) AS DECIMAL(10,3)) AS avg_interest_rate_MoM_change_pct
        FROM
            monthly_avg_interest AS m1
        LEFT JOIN
            monthly_avg_interest AS m2
        ON
            m1.month_num = m2.month_num + 1
        ORDER BY
            m1.month_num;
        ```
    *   **Result:**

        ![image.png](../src/img/SQL/KPI/4.%20Average%20Interest%20Rate/avg_intestest_rate_mom.png)

#### **5. Average Debt-to-Income Ratio (DTI):**

**Explanation:** This KPI helps assess the overall financial health of the bank's borrowers by calculating the average debt-to-income (DTI) ratio across all loans. A high DTI ratio indicates borrowers who may be overextended and potentially at risk of defaulting on their loans.

**Components:**

1.  **Overall Average Debt-to-Income (DTI) Ratio**
    *   **Explanation:** This metric calculates the average DTI ratio across all loans within the dataset. It's important to note that DTI calculations typically require borrower income information, which might not be available in your dataset. 
    *   **Query:**
        ```sql
        SELECT
            ROUND(AVG(dti), 3) AS avg_dti
        FROM
            bank_loan_data
        ```
    *   **Result:**
  
        ![image.png](../src/img/SQL/KPI/5.%20Average%20Debt%20to%20Income%20Ratio/avg_dti_total.png)

2.  **Monthly Average Debt-to-Income (DTI) Ratio**
    *   **Explanation:** This metric tracks the average DTI ratio on a month-by-month basis, revealing potential trends in borrower risk profiles over time.
    *   **Query:**
        ```sql
        SELECT
            DATENAME(month, issue_date) AS month,
            ROUND(AVG(dti), 3) AS avg_dti
        FROM
            bank_loan_data
        GROUP BY
            DATENAME(month, issue_date),
            MONTH(issue_date)
        ORDER BY
            MONTH(issue_date);
        ```
    *   **Result:**

        ![image.png](../src/img/SQL/KPI/5.%20Average%20Debt%20to%20Income%20Ratio/avg_dti_mtd.png)

    *   **Note:** To simplify the next calculation, I've stored the revised query results in a new table called "[table name]".
    *   **Query:**
        ```sql
        SELECT
            DATENAME(month, issue_date) AS month,
            MONTH(issue_date) AS month_num,
            ROUND(AVG(dti), 3) AS avg_dti
        INTO
            monthly_avg_dti
        FROM
            bank_loan_data
        GROUP BY
            DATENAME(month, issue_date),
            MONTH(issue_date)
        ORDER BY
            MONTH(issue_date);
        ```

3.  **Month-Over-Month (MoM) Percentage Change in Average DTI**
    *   **Explanation:** This metric calculates the percentage change in the average DTI ratio from one month to the next, highlighting shifts in the overall risk profile of borrowers over time.
    *   **Query:**
        ```sql
        SELECT
            m1.month AS Month,
            CAST(((m1.avg_dti - m2.avg_dti) * 100.0 / m2.avg_dti) AS DECIMAL(10,3)) AS avg_dti_MoM_change_pct
        FROM
            monthly_avg_dti AS m1
        LEFT JOIN
            monthly_avg_dti AS m2
        ON
            m1.month_num = m2.month_num + 1
        ORDER BY
            m1.month_num;
        ```
    *   **Result:**

        ![image.png](../src/img/SQL/KPI/5.%20Average%20Debt%20to%20Income%20Ratio/avg_dti_mom.png)

#### **6. Good Loan vs. Bad Loan Performance:**

**Explanation:** Evaluating the performance of your loan portfolio requires understanding the health of both "good loans" and "bad loans." This KPI uses loan status criteria to distinguish between these categories and provides key metrics for assessing lending effectiveness and potential risk areas.

**Good Loans:** A loan is considered "good" if its status falls under either "Fully Paid" or "Current."

**Bad Loans:** A loan is considered "bad" if its status is "Charged Off."

**Components:**

##### **6.1. Good Loan KPIs:**

1. **Good Loan Application Percentage:**
    * **Explanation:** This metric calculates the percentage of loan applications classified as good loans. 
    * **Query:**
        ```sql
        SELECT
            CAST(SUM(IIF(loan_status = 'Fully Paid' OR loan_status = 'Current', 1.0, 0.0))/COUNT(*) * 100 AS DECIMAL(10,3)) AS Good_Loan_App_Pct
        FROM
            bank_loan_data;
        ```
    * **Result:** 
    
        ![image.png](../src/img/SQL/KPI/6.%20Good%20Loan%20vs.%20Bad%20Loan%20Performance/Good%20Loan/goodloanpct.png)

2. **Good Loan Applications:**
    * **Explanation:** This metric identifies the total number of loan applications that fall under the good loan category.
    * **Query:**
        ```sql
        SELECT
            COUNT(*) AS total_num_of_good_loans
        FROM
            bank_loan_data
        WHERE
            loan_status = 'Fully Paid' OR loan_status = 'Current';
        ```
    * **Result:**

        ![image.png](../src/img/SQL/KPI/6.%20Good%20Loan%20vs.%20Bad%20Loan%20Performance/Good%20Loan/totalamtgoodloan.png)

3. **Good Loan Funded Amount:**
    * **Explanation:** This metric calculates the total amount of funds disbursed for good loans. 
    * **Query:**
        ```sql
        SELECT
            SUM(loan_amount) AS gl_total_funded_amount
        FROM
            bank_loan_data
        WHERE
            loan_status = 'Fully Paid' OR loan_status = 'Current';
        ```
    * **Result:**

        ![image.png](../src/img/SQL/KPI/6.%20Good%20Loan%20vs.%20Bad%20Loan%20Performance/Good%20Loan/totalfundedamount.png)

4. **Good Loan Total Received Amount:**
    * **Explanation:** This metric tracks the total amount of money received from borrowers for good loans.
    * **Query:**
        ```sql
        SELECT
            SUM(total_payment) AS gl_total_amount_received
        FROM
            bank_loan_data
        WHERE
            loan_status = 'Fully Paid' OR loan_status = 'Current';
        ```
    * **Result:**

        ![image.png](../src/img/SQL/KPI/6.%20Good%20Loan%20vs.%20Bad%20Loan%20Performance/Good%20Loan/totalamtreciveidgl.png)

##### **6.2. Bad Loan KPIs:**

1. **Bad Loan Application Percentage:**
    * **Explanation:** This metric calculates the percentage of loan applications classified as bad loans.
    * **Query:**
        ```sql
        SELECT
            CAST(SUM(IIF(loan_status = 'Charged Off', 1.0, 0.0))/COUNT(*) * 100 AS DECIMAL(10,3)) AS Bad_Loan_App_Pct
        FROM
            bank_loan_data;
        ```
    * **Result:**

        ![image.png](../src/img/SQL/KPI/6.%20Good%20Loan%20vs.%20Bad%20Loan%20Performance/Bad%20Loan/badloanpct.png)

2. **Bad Loan Applications:**
    * **Explanation:** This metric identifies the total number of loan applications that fall under the bad loan category.
    * **Query:**
        ```sql
        SELECT
            COUNT(*) AS total_num_of_bl
        FROM
            bank_loan_data
        WHERE
            loan_status = 'Charged Off';
        ```
    * **Result:**

        ![image.png](../src/img/SQL/KPI/6.%20Good%20Loan%20vs.%20Bad%20Loan%20Performance/Bad%20Loan/bl_total_amt.png)      

3. **Bad Loan Funded Amount:**
    * **Explanation:** This metric calculates the total amount of funds disbursed for bad loans.
    * **Query:**
        ```sql
        SELECT
            SUM(loan_amount) AS bl_total_funded_amount
        FROM
            bank_loan_data
        WHERE
            loan_status = 'Charged Off';
        ```
    * **Result:** 

        ![image.png](../src/img/SQL/KPI/6.%20Good%20Loan%20vs.%20Bad%20Loan%20Performance/Bad%20Loan/bl_funded_amt.png)

4. **Bad Loan Total Received Amount:**
    * **Explanation:** This metric tracks the total amount of money received from borrowers for bad loans.
    * **Query:**
        ```sql
        SELECT
            SUM(total_payment) AS bl_total_amount_received
        FROM
            bank_loan_data
        WHERE
            loan_status = 'Charged Off';
        ```
    * **Result:**

        ![image.png](../src/img/SQL/KPI/6.%20Good%20Loan%20vs.%20Bad%20Loan%20Performance/Bad%20Loan/bl_total_amt_recvied.png)

#### **7. Loan Status Grid View**

**Explanation:** This report offers a comprehensive summary of key lending performance indicators, categorized by 'Loan Status.' This structured view allows the bank to quickly assess the health of different segments within its loan portfolio, compare KPIs across loan statuses, and potentially identify areas for risk mitigation or targeted growth opportunities.  
* **Query:**
    ```sql
    SELECT
        loan_status,
        COUNT(*) AS total_loan_app,
        CAST((CAST(COUNT(*)  AS FLOAT)/ CAST(38576 AS FLOAT) * 100) AS DECIMAL(10,3)) AS total_loan_app_pct, 
        SUM(loan_amount) AS total_funded_amt,
        SUM(total_payment) AS total_amt_received,
        CAST(((CAST(SUM(total_payment) AS FLOAT) - CAST(SUM(loan_amount) AS FLOAT)) / CAST(SUM(loan_amount) AS FLOAT) * 100) AS DECIMAL(10,3)) AS loan_profit_ratio,
        SUM(IIF(MONTH(issue_date) = 12, loan_amount, 0)) AS mtd_funded_amt,
        SUM(IIF(MONTH(issue_date) = 12, total_payment, 0)) AS mtd_amt_received,
        ROUND(AVG(int_rate) * 100, 3) AS avg_interest_rate,
        ROUND(AVG(dti), 3) AS avg_dti
    FROM
        bank_loan_data
    GROUP BY
        loan_status;
    ```
* **Result:**

    ![image-2.png](../src/img/SQL/KPI/7.%20Loan%20Status%20Grid%20View/grid_view.png)

## **Creating SQL Tables for Tableau Overview**

Beyond standard KPIs, we aim to visually represent critical loan-related metrics and trends through a variety of charts in our Tableau dashboard. These visualizations will offer a clear and intuitive overview of our lending operations, allowing stakeholders to easily grasp insights and make data-driven decisions.

**Building the Foundation: SQL Queries for Dashboard Tables**

Before diving into the creation of these charts, we first need to prepare the data in a format that is readily consumable by Tableau. To achieve this, we'll utilize SQL queries to extract and aggregate specific information from our database, generating structured tables that will serve as the backbone for our dashboard visualizations. Each query will be designed with the goal of answering specific business questions and highlighting key relationships within the loan data.

This section will detail the SQL queries used to generate each table, along with the rationale behind the chosen metrics and aggregation methods.


#### **1. Monthly Trends by Issue Date (Line Chart):**

* **Explanation:** This query aggregates key loan metrics (total loan applications, total funded amount, and total amount received) by the month in which the loan was issued. This table will provide a foundational dataset for creating visualizations that showcase monthly trends in these metrics within the Tableau dashboard.

* **Query:**
  ```sql
  SELECT
    DATENAME(month, issue_date) AS Month,
    COUNT(*) AS total_loan_application,
    SUM(loan_amount) AS total_funded_amount,
    SUM(total_payment) AS total_amount_received
  FROM 
    bank_loan_data
  GROUP BY
    DATENAME(month, issue_date),
    MONTH(issue_date)
  ORDER BY
    MONTH(issue_date);
  ```

* **Result:**
    
    ![image.png](../src/img/SQL/Dashboard_Overview/1_monthly_trends_by_isssue.png)

#### **2. Regional Analysis by State (Filled Map):**

**Explanation:** This query aims to provide insights into loan performance across different geographical regions (states). By aggregating loan metrics such as total applications, total funded amount, and total amount received by state, we can identify potential regional trends and disparities in lending activity.

* **Query:**
  ```sql
  SELECT
    address_state AS state_name,
    COUNT(*) AS total_loan_application,
    SUM(loan_amount) AS total_funded_amount,
    SUM(total_payment) AS total_amount_received
  FROM 
    bank_loan_data
  GROUP BY
    address_state
  ORDER BY
      address_state;
  ```

* **Result:** 

    ![image.png](../src/img/SQL/Dashboard_Overview/2_regional_analysis_by_state.png)


#### **3. Loan Term Analysis (Donut Chart):**

**Explanation:** This query delves into the distribution of loan terms within the portfolio. By aggregating key metrics based on loan term length, we can gain insights into the popularity of different terms, their associated funding amounts, and repayment patterns. This data will be visualized in the Tableau dashboard as a donut chart, providing a clear visual breakdown of loan terms.

* **Query:**
 
  ```sql
  SELECT
    term AS loan_term_length,
    COUNT(*) AS total_loan_application,
    SUM(loan_amount) AS total_funded_amount,
    SUM(total_payment) AS total_amount_received
  FROM 
	bank_loan_data
  GROUP BY
	term
  ORDER BY
	term;
  ```

* **Result:** 

	![image.png](../src/img/SQL/Dashboard_Overview/3_loan_term_analysis.png)

#### **4. Employee Length Analysis (Bar Chart):**

**Explanation:** This query analyzes loan metrics in relation to the employment length of borrowers.  It categorizes borrowers based on their years of employment and aggregates key metrics like total loan applications, total funded amount, and total amount received. The results will be visualized in the Tableau dashboard as a bar graph, comparing loan activity and repayment behavior across different employment length categories.

* **Query:**

  ```sql
  SELECT
    emp_length AS employee_len,
    COUNT(*) AS total_loan_application,
    SUM(loan_amount) AS total_funded_amount,
    SUM(total_payment) AS total_amount_received
  FROM 
    bank_loan_data
  GROUP BY
    emp_length
  ORDER BY
    emp_length;
  ```

* **Result:** 

	![image.png](../src/img/SQL/Dashboard_Overview/4_employee_length_analysis.png)

#### **5. Loan Purpose Breakdown (Bar Chart):**

**Explanation:** This query aims to shed light on the different purposes for which borrowers are seeking loans. By aggregating key metrics based on loan purpose, we can gain insights into the most common reasons for borrowing, the average loan amounts associated with each purpose, and their relative success rates. This information will be presented in a bar chart in the Tableau dashboard, allowing stakeholders to visually compare these metrics across various loan purposes.

* **Query:**
	```sql
	SELECT
	  purpose AS loan_purpose,
	  COUNT(*) AS total_loan_application,
	  SUM(loan_amount) AS total_funded_amount,
	  SUM(total_payment) AS total_amount_received
	FROM 
	  bank_loan_data
	GROUP BY
	  purpose
	ORDER BY
	  purpose;
	```

* **Result:**
    
	
	![image.png](../src/img/SQL/Dashboard_Overview/5_loan_purpose_analysis.png)

#### **6. Home Ownership Analysis (Tree Map):**

**Explanation:** This query aims to examine the relationship between home ownership status and various loan metrics. By aggregating data based on whether borrowers own, rent, or mortgage their homes, we can uncover potential insights into borrower behavior, risk profiles, and the impact of home ownership on loan performance. This information will be presented as a tree map in the Tableau dashboard, providing a hierarchical view of loan metrics categorized by different home ownership statuses.

* **Query:**

  ```sql
  SELECT
	home_ownership,
	COUNT(*) AS total_loan_application,
	SUM(loan_amount) AS total_funded_amount,
	SUM(total_payment) AS total_amount_received
  FROM 
	bank_loan_data
  GROUP BY
	home_ownership
  ORDER BY
	home_ownership;
  ```

* **Result:**

    ![image.png](../src/img/SQL/Dashboard_Overview/6_home_ownership_analysis.png)

## **Tableau Dashboard**

As previously stated, I have transformed the loan dataset into an interactive Tableau dashboard, providing a comprehensive platform for analyzing and exploring the lending landscape of 2021. The following sections will guide you through the various components of the dashboard, explaining its structure, features, and the insights it offers.

**To access the Tableau dashboard, please click** **[here](https://public.tableau.com/views/BankLoanReportDashboard_17165926472590/Summary?:language=en-US&publish=yes&:sid=&:display_count=n&:origin=viz_share_link).**

Once you have accessed the dashboard, you will find it organized into three main sections, each designed to provide a unique perspective on the data:

1.  **Key Performance Indicators (KPIs):**  Every section presents a consistent set of metrics across all views, providing a high-level overview of loan activity.

2.  **Summary:**  This section offers a comprehensive summary of loan performance and outcomes.

3.  **Overview:**  This section provides a comparative analysis of lending activity across various dimensions.

4.  **Details:**  This section offers a granular view of individual loan records, allowing for in-depth exploration and analysis.

Each section is equipped with custom filters, enabling you to tailor the analysis to your specific interests and explore the data in a way that is most relevant to your needs.

Now, let's delve into each section in more detail:

**Key Performance Indicators (KPIs)**

![KPI.png](../src/img/Dashboard/KPI.png)

Each section of the dashboard (Summary, Overview, and Details) features a consistent set of KPIs, providing a standardized reference point for users to track key metrics across the different analyses. These KPIs include:

* **Total Loan Applications:** The total number of loan applications in the dataset.
* **Total Funded Amount:** The total amount of money disbursed through approved loans.
* **Total Amount Received:** The total amount of money received from borrowers through loan repayments.
* **Average Interest Rate:** The average interest rate applied to loans in the dataset.
* **Average Debt-to-Income (DTI) Ratio:** The average ratio of borrowers' total debt payments to their gross income.

Within this KPI section, users can find additional breakdowns of these metrics:

* **Average Monthly Values:** Average values for each KPI are calculated on a monthly basis.
* **Month-to-Month Changes:** Average percentage changes in each KPI from one month to the next.

These KPIs serve as a foundation for understanding overall lending activity and trends within the dataset, and their presence in each section allows for easy comparison and analysis across different views of the data.

**Summary Section**

![Summary.png](../src/img/Dashboard/Summary.png)

The Summary section provides a high-level overview of the KPIs and delves deeper into the performance and status of loans. Key elements include:

* **Loan Performance Breakdown:** Comparison of "good" loans (paid off or current) and "bad" loans (charged off or defaulted), including the number and total amounts for each category.
* **Loan Status Table:** Detailed breakdown of loan statuses, showing the distribution of loans across various stages and the corresponding total amounts.

This section offers a comprehensive summary of the overall health of the loan portfolio and allows users to quickly grasp the distribution of loan outcomes.

**Overview Section**

![Overview.png](../src/img/Dashboard/Overview.png)

The Overview section provides a comparative analysis of lending activity across various dimensions. It leverages the same KPIs as the Summary section but presents them in a more granular manner, enabling users to explore trends and patterns within specific subsets of the data. Key features include:

* **Monthly Trend Analysis:** Visualizations showcasing the monthly evolution of loan applications, funded amounts, and amounts received.
* **Geographic Comparison:** Breakdown of loan metrics by state.
* **Borrower Characteristic Comparison:** Analysis of loan metrics by employment length, loan purpose, and homeownership status.

Users can interact with this section through **custom filters** specific to this view, allowing them to focus on particular months, states, borrower characteristics, or loan purposes.

**Details Section**

![Details.png](../src/img/Dashboard/Details.png)

The Details section offers a comprehensive and granular view of the loan dataset through a tabular format. This section provides users with the ability to examine individual loan records in detail, offering a wealth of information for further analysis and exploration.

Key features of the Details section include:

* **Comprehensive Loan Data:** Presentation of all relevant loan information, including borrower details, loan specifics, repayment status, and other key variables.
* **Custom Filters:** Users can apply filters specific to this view to narrow down the displayed data and focus on specific loan characteristics or borrower attributes.
* **Changeable Measures:** The table allows users to select different measures (e.g., loan amount, interest rate, DTI ratio) to be displayed, providing flexibility in how they analyze the loan data.

The Details section empowers users to delve deep into the data, offering the ability to scrutinize individual loans and identify patterns or outliers that may not be apparent in the higher-level summaries or visualizations provided in the other sections. This granular view can be particularly valuable for researchers, analysts, or lenders seeking to understand the nuances of individual loan performance and its contributing factors.

**Conclusion of Dashboard Section**

By combining these sections with their respective KPIs and custom filters, the Tableau dashboard offers a flexible and comprehensive tool for exploring and analyzing the 2021 loan dataset. Whether users seek high-level summaries, comparative analyses, or granular details, the dashboard caters to a wide range of analytical needs and provides valuable insights into the lending landscape of that year.

## **Detailed Analysis**

### **Leveraging Python for Real-Time Exploration**

To complement the insights gained from the initial SQL and Tableau analyses, and to facilitate a more dynamic and granular exploration of the loan dataset, we will leverage the power and flexibility of Python within the Jupyter Notebook environment. This approach offers several key advantages:

* **Real-time interaction:** Python allows for immediate execution of code and visualization of results, enabling us to rapidly iterate on our analysis and test new hypotheses.
* **Customizable analysis:** Python provides a vast ecosystem of libraries and tools, enabling us to tailor our analysis to the specific questions and nuances of the loan data.
* **Data manipulation and transformation:** Python's Pandas library offers powerful capabilities for cleaning, reshaping, and transforming the data to suit our analytical needs.

By transitioning to Python for this detailed analysis phase, we gain the ability to delve deeper into the data, uncover hidden patterns, and explore complex relationships that may not have been readily apparent in the previous stages of analysis. 

### **In-Depth Exploration of Loan Data**


Before embarking on our in-depth exploration of the loan data, let's first initialize our Python environment by loading the necessary libraries and importing the dataset:

In [1]:
# Import necessary libraries
import pandas as pd
import numpy as np

# Load the financial loan data from a CSV file into a DataFrame
df = pd.read_csv('../src/financial_loan.csv')

With these initial steps complete, we are now prepared to delve deeper into the intricacies of the loan data and uncover its hidden insights.

In [2]:
# Import necessary libraries
import pandas as pd
import numpy as np

# Load the financial loan data from a CSV file into a DataFrame
df = pd.read_csv('../src/financial_loan.csv')

To delve deeper into the financial implications of loan performance, we will utilize Python to examine the profit margins associated with "good" loans (current and fully paid) and the losses incurred from "bad" loans (charged off). By quantifying these figures, we can gain a clearer understanding of the overall profitability of the loan portfolio and identify potential areas for risk mitigation.

Let's begin by filtering the loan dataset to isolate the relevant loan statuses and calculate the corresponding profit margins.

#### **Overview and Analysis of Loan Portfolio**

In [3]:
# Filter for good loans and calculate the total loan amount and total payment by status
good_loan = df[(df['loan_status'] == 'Fully Paid') | (df['loan_status'] == 'Current')] \
                    .groupby('loan_status')[['loan_amount', 'total_payment']].sum()

# Calculate profit for each loan
good_loan['profit'] = good_loan['total_payment'] - good_loan['loan_amount']

# Calculate profit margin as a percentage for each loan
good_loan['profit_margin'] = ((good_loan['total_payment'] - good_loan['loan_amount']) / 
                              good_loan['loan_amount']) * 100

# Calculate total values for 'Fully Paid' and 'Current' loans
total_row = good_loan[['loan_amount', 'total_payment', 'profit']].sum()

# Create a new DataFrame for the combined totals row
total_df = pd.DataFrame([total_row], index=['Combined'])

# Calculate profit margin for the combined row
total_df['profit_margin'] = (total_df['total_payment'] - total_df['loan_amount']) / total_df['loan_amount'] * 100

# Append the combined totals row to the original DataFrame using concat
good_loan = pd.concat([good_loan, total_df])

# Ensure all columns are properly formatted (if necessary)
good_loan[['loan_amount', 'total_payment', 'profit', 'profit_margin']] = good_loan[['loan_amount', 'total_payment', 'profit', 'profit_margin']].apply(pd.to_numeric, errors='ignore')

C:\Users\AAA\AppData\Local\Temp\ipykernel_50284\453163596.py:25: FutureWarning: errors='ignore' is deprecated and will raise in a future version. Use to_numeric without passing `errors` and catch exceptions explicitly instead
  good_loan[['loan_amount', 'total_payment', 'profit', 'profit_margin']] = good_loan[['loan_amount', 'total_payment', 'profit', 'profit_margin']].apply(pd.to_numeric, errors='ignore')


In [4]:
# Display the updated good_loan DataFrame
good_loan

loan_amount  total_payment    profit  profit_margin
Current        18866500       24199914   5333414      28.269229
Fully Paid    351358350      411586256  60227906      17.141447
Combined      370224850      435786170  65561320      17.708514

In [5]:
# Filter for bad loans and calculate the total loan amount and total payment by status
bad_loan = df[(df['loan_status'] == 'Charged Off')] \
                    .groupby('loan_status')[['loan_amount', 'total_payment']].sum()

# Calculate profit for each loan
bad_loan['profit'] = bad_loan['total_payment'] - bad_loan['loan_amount']

# Calculate profit margin as a percentage for each loan
bad_loan['profit_margin'] = ((bad_loan['total_payment'] - bad_loan['loan_amount']) / 
                              bad_loan['loan_amount']) * 100

In [6]:
# Display the updated bad_loan DataFrame
bad_loan

loan_amount  total_payment    profit  profit_margin
loan_status                                                     
Charged Off     65532225       37284763 -28247462     -43.104689

In [7]:
# Calculate total profit by subtracting the total loan amount from the total payments
profit = sum(df['total_payment']) - sum(df['loan_amount'])

# Calculate profit margin as a percentage of total loan amount
profit_margin = ((sum(df['total_payment']) - sum(df['loan_amount'])) / 
                 sum(df['loan_amount'])) * 100

# Print formatted results for profit and profit margin
print(f"Total Profit: ${profit:,.2f}, Profit Margin: {profit_margin:.2f}%")

Total Profit: $37,313,858.00, Profit Margin: 8.56%


Analysis of profit margins across loan categories reveals a significant disparity between the performance of "good" loans and "bad" loans. While current loans boast a 28.27% profit margin and fully paid loans achieve a respectable 17.14%, combining these into a general "good loan" category results in a 17.71% profit margin. This indicates a healthy return on investment for the bank when loans are successfully repaid.

However, the picture is starkly different for "bad" loans (charged off), which suffer a substantial 43.1% loss. This indicates that for every dollar lent out in these loans, the bank loses 43 cents. Although the overall profit margin across all loans remains positive at 8.56%, the significant losses incurred from bad loans raise concerns about the sustainability and potential for optimization within the loan portfolio.

It's crucial to contextualize these figures within the unique economic landscape of 2021. The economy was still grappling with the aftermath of the COVID-19 pandemic, which had disrupted businesses, employment, and consumer spending. Government intervention, including the promotion of low interest rates, aimed to stimulate economic activity, but these efforts also created an environment where the risk of loan defaults might have been elevated.

Without comparative data from previous years, it's difficult to definitively assess the performance of the bank's loan portfolio in 2021. However, the substantial losses associated with bad loans warrant further investigation into the underlying factors contributing to these negative outcomes.

Moving forward, our analysis will delve deeper into the dataset to identify potential areas for improvement. By exploring relationships between loan purpose, borrower characteristics, and other relevant factors, we aim to pinpoint strategies for minimizing losses on bad loans while maximizing the potential of good loans. Ultimately, this analysis will inform decision-making and contribute to the development of more resilient and profitable lending practices in the future.

#### **Loan Application Analysis**

In examining the Tableau dashboard, specifically the "Total Loan Application by Purpose" visualization, a compelling insight emerged: a significant majority of borrowers sought loans for debt consolidation. This was evident in the stark contrast between the 18.2K loans taken for debt consolidation and the 5K loans for the second most common purpose, credit card refinancing.

![total_purpose.png](../src/img/Dashboard/total_purpose.png)

This disproportionate prevalence of debt consolidation loans piqued my curiosity and prompted a deeper investigation into the underlying motivations behind this trend. Why were so many individuals seeking to consolidate their existing debts through new loans? What were the financial implications for both borrowers and the lending institutions?

To delve into these questions, I decided to conduct a more comprehensive analysis of loan purposes and their associated outcomes. My goal was to assess the profitability of loans issued for different purposes, considering factors such as loan amounts, interest rates, and repayment behaviour.

In the following sections, I will utilize Python to dissect the data and uncover insights into the relationship between loan purpose, profitability, and borrower characteristics. By exploring these connections, we can gain a better understanding of the factors driving debt consolidation and its impact on the lending landscape in 2021.

#### **Profitability Across Loan Purposes**

In [8]:
# Group the data by 'purpose' and aggregate 'id', 'loan_amount', 'total_payment', and 'int_rate'
purpose_breakdown = df.groupby('purpose')[['id', 'loan_amount', 'total_payment', 'int_rate']] \
    .agg({
        'id': 'count',               # Count of IDs per purpose
        'loan_amount': 'sum',        # Sum of loan amounts per purpose
        'total_payment': 'sum',      # Sum of total payments per purpose
        'int_rate': 'mean'           # Average interest rate per purpose
    })

# Calculate profit for each loan purpose
purpose_breakdown['profit'] = purpose_breakdown['total_payment'] - purpose_breakdown['loan_amount']

# Calculate profit margin as a percentage for each loan purpose
purpose_breakdown['profit_margin'] = ((purpose_breakdown['total_payment'] - purpose_breakdown['loan_amount']) / 
                                      purpose_breakdown['loan_amount']) * 100

In [9]:
# Print the DataFrame with the calculated profit and profit margin for each loan purpose
purpose_breakdown

id  loan_amount  total_payment  int_rate    profit  \
purpose                                                                     
Debt consolidation  18214    232459675      253801871  0.125043  21342196   
car                  1497     10223575       11324914  0.105921   1101339   
credit card          4998     58885175       65214084  0.117307   6328909   
educational           315      2161650        2248380  0.116545     86730   
home improvement     2876     33350775       36380930  0.114013   3030155   
house                 366      4824925        5185538  0.123807    360613   
major purchase       2110     17251600       18676927  0.108690   1425327   
medical               667      5533225        5851372  0.115709    318147   
moving                559      3748125        3999899  0.115920    251774   
other                3824     31155750       33289676  0.118585   2133926   
renewable_energy       94       845750         898931  0.114963     53181   
small business       1776     24123100       23814817  0.130295   -308283   
vacation              352      1967950        2116738  0.108753    148788   
wedding               928      9225800       10266856  0.118930   1041056   

                    profit_margin  
purpose                            
Debt consolidation       9.181032  
car                     10.772543  
credit card             10.747882  
educational              4.012213  
home improvement         9.085711  
house                    7.473961  
major purchase           8.261999  
medical                  5.749757  
moving                   6.717332  
other                    6.849220  
renewable_energy         6.288028  
small business          -1.277958  
vacation                 7.560558  
wedding                 11.284181

An analysis of loan purposes and their corresponding profit margins reveals a notable trend: while most loan purposes yielded positive returns for the bank in 2021, small business loans stood out as an exception, experiencing a negative profit margin. This finding is particularly intriguing when considering the economic landscape of the time, as the country was still recovering from the impact of the COVID-19 pandemic.

As shown in the breakdown of loan purposes, all categories except small business loans demonstrated positive profit margins. This suggests that, in general, the bank's lending activities were financially viable during this period. However, the negative profit margin associated with small business loans raises questions about the unique challenges faced by this sector in the post-pandemic recovery phase.

The struggles of small businesses in 2021 can likely be attributed to the lingering effects of COVID-19 restrictions and the slow economic recovery. Many small businesses may have struggled to attract customers, maintain consistent revenue streams, and adapt to the changing economic landscape. These factors could have contributed to an increased risk of loan default and consequently, negative profit margins for the bank.

The lowest profit margin among all loan purposes was observed for small business loans (-1.28%), followed closely by educational loans (4.01%). This stark contrast highlights the vulnerability of small businesses during this period and underscores the importance of understanding the unique challenges they faced.
In light of these findings, further analysis will focus on identifying the specific factors contributing to the high default rates among small business loans. By examining borrower characteristics, loan terms, and other relevant variables, we can gain a deeper understanding of the challenges faced by small businesses in 2021 and potentially identify strategies to mitigate these risks in the future.

Additionally, exploring the reasons behind the relatively low-profit margin for educational loans could shed light on another area where lending practices might be optimized. By identifying commonalities among borrowers who struggle to repay educational loans, lenders can tailor their products and services to better meet the needs of this specific borrower segment.

In [10]:
# Aggregate count of IDs, sum of loan_amount and total_payment, and average interest rate for 'Charged Off' loans by purpose
purpose_breakdown_bl = df[df['loan_status'] == 'Charged Off'].groupby('purpose')[['id', 'loan_amount', 'total_payment', 'int_rate']] \
                        .agg({
                            'id': 'count',                # Count of IDs per purpose
                            'loan_amount': 'sum',         # Sum of loan amounts per purpose
                            'total_payment': 'sum',       # Sum of total payments per purpose
                            'int_rate': 'mean'            # Average interest rate per purpose
                        })

# Calculate profit by subtracting the loan amount from the total payment for each purpose
purpose_breakdown_bl['profit'] = purpose_breakdown_bl['total_payment'] - purpose_breakdown_bl['loan_amount']

# Calculate the profit margin as a percentage for each purpose
purpose_breakdown_bl['profit_margin'] = ((purpose_breakdown_bl['total_payment'] - purpose_breakdown_bl['loan_amount']) /
                                         purpose_breakdown_bl['loan_amount']) * 100


In [11]:
# Print the DataFrame with the calculated profit, profit margin, and average interest rate for each loan purpose
purpose_breakdown_bl.sort_values('profit_margin')

id  loan_amount  total_payment  int_rate    profit  \
purpose                                                                    
moving                84       515750         250822  0.124168   -264928   
small business       455      6668975        3360238  0.142187  -3308737   
medical              100       884850         462652  0.134758   -422198   
home improvement     327      3946725        2092243  0.133302  -1854482   
other                587      4976825        2685685  0.134868  -2291140   
major purchase       206      1914525        1044301  0.129098   -870224   
educational           50       405350         234280  0.127712   -171070   
wedding               86       972850         564925  0.136795   -407925   
Debt consolidation  2651     36257900       21060941  0.143122 -15196959   
renewable_energy      17       155100          93522  0.123947    -61578   
credit card          508      6692450        4050108  0.135150  -2642342   
house                 57       725450         457278  0.153293   -268172   
vacation              50       285350         186896  0.116118    -98454   
car                  155      1130125         740872  0.124629   -389253   

                    profit_margin  
purpose                            
moving                 -51.367523  
small business         -49.613876  
medical                -47.714076  
home improvement       -46.987870  
other                  -46.036178  
major purchase         -45.453781  
educational            -42.203034  
wedding                -41.930925  
Debt consolidation     -41.913511  
renewable_energy       -39.702128  
credit card            -39.482432  
house                  -36.966297  
vacation               -34.502891  
car                    -34.443358

A closer examination of the relationship between loan purpose and loan outcomes has revealed a potential connection between interest rates and loan defaults. While analyzing the breakdown of good versus bad loans, I observed a discrepancy in the average interest rates for these categories. This sparked my curiosity regarding the potential impact of interest rates on the likelihood of loan default.

Specifically, it appears that the average interest rate for charged-off (defaulted) loans might be higher than that of current or fully paid loans. This suggests a possible correlation between higher interest rates and increased difficulty in repayment, leading to default.

Several factors could contribute to this phenomenon:

* **Timing of Loan Origination:** Loans taken out earlier in the year, when interest rates may have been higher, might be more likely to default than those taken out later when rates potentially decreased.
* **Borrower Location:** Interest rates could vary based on the borrower's geographic location, with certain regions facing higher rates and subsequently a greater risk of default.
* **Individual Borrower Circumstances:** Borrowers with lower credit scores or less stable financial situations might have qualified for loans only at higher interest rates, making repayment more challenging.

Before drawing definitive conclusions about the relationship between loan purpose and default rates, it's crucial to investigate the influence of interest rates. By analyzing the distribution of interest rates across different loan categories and examining the correlation between interest rates and default status, we can gain a more nuanced understanding of the factors contributing to loan repayment success or failure.

This analysis will help us determine whether the observed differences in default rates are solely attributable to loan purposes or if interest rates play a significant role. If interest rates prove to be a major factor, this insight could inform lending practices and potentially lead to adjustments in interest rate policies to minimize defaults and maximize the success of future loans.

#### **Interest Rates and Loan Outcomes**

In [12]:
# Pivot the table to compare interest rates of each purpose for different loan statuses
int_comp = df.pivot_table(
    index='purpose', 
    columns='loan_status', 
    values='int_rate', 
    aggfunc='mean'
)

In [13]:
# Print the pivoted table
int_comp

loan_status         Charged Off   Current  Fully Paid
purpose                                              
Debt consolidation     0.143122  0.153650    0.120765
car                    0.124629  0.127694    0.102853
credit card            0.135150  0.155331    0.114367
educational            0.127712       NaN    0.114438
home improvement       0.133302  0.147253    0.110172
house                  0.153293  0.139207    0.117379
major purchase         0.129098  0.137829    0.105895
medical                0.134758  0.148883    0.111559
moving                 0.124168  0.141843    0.114051
other                  0.134868  0.147996    0.114387
renewable_energy       0.123947  0.114900    0.112954
small business         0.142187  0.156651    0.124418
vacation               0.116118  0.137975    0.107125
wedding                0.136795  0.165224    0.115874

Upon analyzing the interest rates across different loan purposes and statuses, a fascinating pattern emerges. It appears that fully paid-off loans generally have lower interest rates compared to charged-off loans, which tend to have higher rates. However, an intriguing observation is that current loans often exhibit interest rates that are, in some categories, even higher than those of charged-off loans.

Several possible explanations can be offered for this phenomenon:

* **Borrower Profile:** It's likely that borrowers who fully paid off their loans had stronger financial profiles, including higher credit scores, stable employment, and more assets. This would have qualified them for lower interest rates and enabled them to manage repayment more effectively. Conversely, those with charged-off loans might have had less favourable financial circumstances, leading to higher interest rates and ultimately default.

* **Loan Timing:** The timing of loan origination could also play a role. If interest rates fluctuated throughout 2021, loans taken out earlier in the year may have had higher rates than those issued later. This could explain why some current loans, even if they are being repaid responsibly, have higher rates than charged-off loans originated earlier in the year.

* **Loan Terms:** The length of the loan term could also influence interest rates. Longer-term loans often come with higher interest rates due to the increased risk associated with extended repayment periods. Borrowers who opted for longer terms to make monthly payments more manageable might have ended up with higher rates, even if they are currently keeping up with their payments.

* **Economic Context:** The unique economic landscape of 2021, characterized by a recovering economy and fluctuating interest rates, likely contributed to the observed variations in interest rates across loan statuses. The demand for loans, coupled with the uncertainty surrounding economic recovery, could have led to an increase in interest rates later in the year, affecting those with current loans.

The fact that current loans sometimes exhibit higher interest rates than charged-off loans raises questions about the sustainability of these loans. Are borrowers with current loans able to manage the higher interest payments, or could they potentially face difficulties in the future? Are there specific loan purposes where this pattern is more pronounced, suggesting a need for closer scrutiny or potential policy adjustments?

To address these questions, further analysis is needed. By examining the interaction between interest rates, loan terms, borrower characteristics, and economic conditions in 2021, we can gain a deeper understanding of the factors influencing loan repayment success and develop strategies to mitigate the risk of future defaults.

#### **Statewise Loan Performance**

In [14]:
# Aggregate count of loans, mean DTI and interest rate, and sum of loan amounts and total payments by state
state_breakdown = df.groupby('address_state')[['id', 'dti', 'int_rate', 'loan_amount', 'total_payment']] \
                    .agg({
                        'id': 'count',  # Count of loans per state
                        'dti': 'mean',  # Average debt-to-income ratio
                        'int_rate': 'mean',  # Average interest rate
                        'loan_amount': 'sum',  # Total loan amount
                        'total_payment': 'sum'  # Total payments
                    })

# Calculate profit by subtracting the total loan amount from the total payments for each state
state_breakdown['profit'] = state_breakdown['total_payment'] - state_breakdown['loan_amount']

# Calculate the profit margin as a percentage for each state
state_breakdown['profit_margin'] = ((state_breakdown['total_payment'] - state_breakdown['loan_amount']) /
                                    state_breakdown['loan_amount']) * 100

In [15]:
# Print the DataFrame with the aggregated data and calculated profit metrics for each state
state_breakdown.sort_values('id', ascending=False)

id       dti  int_rate  loan_amount  total_payment   profit  \
address_state                                                                  
CA             6894  0.128443  0.121482     78484125       83901234  5417109   
NY             3701  0.122125  0.121143     42077050       46108181  4031131   
FL             2773  0.135271  0.119781     30046125       31601905  1555780   
TX             2664  0.139467  0.120470     31236650       34392715  3156065   
NJ             1822  0.128259  0.122351     21657475       23425159  1767684   
IL             1486  0.133111  0.120541     17124225       18875941  1751716   
PA             1482  0.138422  0.116906     15826525       17462908  1636383   
VA             1375  0.134691  0.122175     15982650       17711443  1728793   
GA             1355  0.139886  0.119620     15480325       16728040  1247715   
MA             1310  0.126445  0.118715     15051000       16676279  1625279   
OH             1188  0.145976  0.120245     12991375       14330148  1338773   
MD             1027  0.129233  0.124916     11911400       12985170  1073770   
AZ              833  0.131855  0.122431      9206000       10041986   835986   
WA              805  0.130510  0.122631      8855525        9531739   676214   
CO              770  0.136090  0.118150      8976000        9845810   869810   
NC              759  0.137806  0.121221      8787575        9534813   747238   
CT              730  0.128976  0.119101      8435575        9357612   922037   
MI              685  0.134616  0.120066      7829900        8543660   713760   
MO              660  0.144717  0.117236      7151175        7692732   541557   
MN              592  0.135916  0.116876      6302600        6750746   448146   
NV              482  0.135447  0.125404      5307375        5451443   144068   
SC              464  0.137935  0.118348      5080475        5462458   381983   
WI              446  0.134770  0.120678      5070450        5485161   414711   
OR              436  0.137516  0.119970      4720150        4966903   246753   
AL              432  0.141753  0.118594      4949225        5492272   543047   
LA              426  0.142009  0.117323      4498900        5001160   502260   
KY              320  0.140750  0.118976      3504100        3792530   288430   
OK              293  0.141418  0.118953      3365725        3712649   346924   
KS              260  0.141783  0.118186      2872325        3247394   375069   
UT              252  0.132625  0.119757      2849225        2952412   103187   
AR              236  0.152133  0.117557      2529700        2777875   248175   
DC              214  0.125146  0.120467      2652350        2921854   269504   
RI              196  0.127242  0.118652      1883025        2001774   118749   
NM              183  0.138801  0.116613      1916775        2084485   167710   
HI              170  0.140616  0.125938      1850525        2080184   229659   
WV              167  0.148839  0.117253      1830525        1991936   161411   
NH              161  0.141463  0.117891      1917900        2101386   183486   
DE              110  0.141715  0.119956      1138100        1269136   131036   
MT               79  0.135662  0.120765       829525         892047    62522   
WY               79  0.143459  0.125241       890750        1046050   155300   
AK               78  0.150435  0.127915      1031800        1108570    76770   
SD               63  0.148140  0.114314       606150         656514    50364   
VT               54  0.135194  0.112231       504100         534973    30873   
MS               19  0.149953  0.117900       139125         149342    10217   
TN               17  0.115918  0.105612       162175         141522   -20653   
IN                9  0.159711  0.108011        86225          85521     -704   
ID                6  0.145367  0.115117        59750          65329     5579   
NE                5  0.162820  0.118140        31700          24542    -7158   
IA               

Analyzing the geographical distribution of loans reveals interesting insights into the varying performance across different states. While California led in terms of loan volume with 6.9K loans and a respectable 6.9% profit margin, New York followed with 3.7K loans and a notably higher profit margin of 9.58%.

However, not all states yielded positive results. Tennessee (TN) and Nebraska (NE) both experienced negative profit margins, despite having relatively low loan volumes (17 and 5 loans respectively). This raises questions about the underlying factors contributing to these losses.

In Nebraska, the average debt-to-income (DTI) ratio was notably higher than in other states. Given the small sample size of 5 loans, it's possible that borrowers in this state faced greater financial strain, making it difficult for them to repay their loans and resulting in negative profit margins for the bank.

Tennessee presents a different scenario, with a DTI ratio not significantly different from other states. However, despite relatively low interest rates, borrowers in this state still defaulted on their loans, resulting in losses for the bank. This suggests that factors other than DTI and interest rates may have played a role in the negative profit margins observed in Tennessee.

While the overall profit margin for loans across all states was positive, the localized losses in Tennessee and Nebraska highlight the importance of understanding regional variations in borrower behaviour and economic conditions. During the 2021 recovery period, the impact of the pandemic likely varied across states, influencing borrowers' financial stability and repayment capacity.

To optimize future lending strategies, the bank should consider several approaches:

* **Enhanced Due Diligence:** In states with negative profit margins or high default rates, the bank may need to conduct more thorough background checks on borrowers, including a deeper assessment of their financial stability, employment status, and credit history.
* **Targeted Interest Rates:** For states with positive profit margins and low loan volumes, the bank could explore lowering interest rates to attract more borrowers and potentially increase overall profitability.
* **Customized Loan Products:** Developing loan products tailored to the specific needs and circumstances of borrowers in different regions could improve repayment outcomes and reduce the risk of default.

By addressing these regional variations and tailoring its lending practices accordingly, the bank can mitigate risk, enhance profitability, and contribute to a more stable and inclusive economic recovery.

Further analysis will delve deeper into the specific factors influencing loan performance in each state. By examining borrower demographics, loan purposes, and economic indicators at a more granular level, we can gain a comprehensive understanding of the lending landscape and identify opportunities for optimization.


## **Conclusion**

The analysis of the 2021 loan dataset has unveiled a multifaceted picture of lending activity during a pivotal period of economic recovery. While the overall profit margin for the bank remained positive, the data revealed significant disparities in loan performance across different purposes and borrower profiles.

Key findings include:

1.  **Debt Consolidation Dominance:** A substantial portion of borrowers sought loans for debt consolidation, highlighting a potential need for financial education and alternative debt management solutions.

2.  **Profitability Disparities:** While most loan purposes generated positive returns, small business loans suffered a negative profit margin, underscoring the unique challenges faced by this sector during the pandemic recovery.

3.  **Interest Rate Influence:** Interest rates appeared to play a role in loan outcomes, with higher rates potentially contributing to increased default rates.

4.  **Regional Variations:** Loan performance varied across states, with some regions experiencing negative profit margins due to factors like higher debt-to-income ratios and potentially varying economic conditions.

Based on these findings, several recommendations can be made:

1.  **Tailored Lending Strategies:** The bank should consider developing specialized loan products and services for small businesses, addressing their unique needs and challenges during periods of economic recovery.

2.  **Interest Rate Optimization:** Careful consideration should be given to interest rate policies, ensuring they are aligned with borrower risk profiles and economic conditions to minimize defaults.

3.  **Enhanced Due Diligence:** In regions with higher default rates, more rigorous borrower assessment and underwriting practices may be necessary to mitigate risk.

4.  **Financial Education Initiatives:** Offering financial literacy programs to borrowers could empower them to make informed decisions and manage their debt more effectively.

This analysis has not only provided a comprehensive overview of the lending landscape in 2021 but also identified key areas for improvement. By implementing these recommendations, the bank can enhance its lending practices, reduce risk, and foster a more sustainable and inclusive financial ecosystem.

Furthermore, this analysis serves as a valuable resource for future research and policy development. By understanding the complex interplay between borrower behaviour, economic conditions, and lending practices, we can create more effective strategies to support both lenders and borrowers during times of economic uncertainty.